unblob/handlers/archive/xiaomi/hdr1.py

```diff
@@ -0,0 +1,142 @@
+    Endian,
+    Extractor,
+    HexString,
+    StructHandler,
+    StructParser,
+    ValidChunk,
+)
+
+logger = get_logger()
+CRC_CONTENT_OFFSET = 12
+SIGNATURE_LEN = 272
+
+# https://lxr.openwrt.org/source/firmware-utils/src/xiaomifw.c
+HDR1_HEADER_DEFINITION = r"""
+    struct xiaomi_header {
+          char magic[4];                /* HDR1 */
+          uint32 signature_offset;
+          uint32 crc32;
+          uint16 unused;
+          uint16 device_id;             /* RA70 */
+          uint32 blob_offsets[8];
+    } hdr_header_t;
+
+    struct xiaomi_blob_header {
+          uint32 magic;                 /* HDR1 */
+          uint32 flash_offset;
+          uint32 size;                  /* Size of blob */
+          uint16 type;                  /* Type of blob */
+          uint16 unused;
+          char name[32];                /* Name of blob */
+    } blob_header_t;
+
+    struct xiaomi_signature_header {
+          uint32 size;
+          uint32 padding[3];
+          uint8 content[0x100];
+    } signature_header_t;
+ """
+
+
+@attrs.define
+class Blob:
+    offset: int
+    name: str
+    size: int
+
+
+class HDRExtractor(Extractor):
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            parser = HDRParser()
+            parser.parse(file)
+            for blob in parser.blobs:
+                chunk = Chunk(
+                    start_offset=blob.offset, end_offset=blob.offset + blob.size
+                )
+                carve_chunk_to_file(outdir.joinpath(Path(blob.name)), file, chunk)
+
+
+class HDRParser:
+    def __init__(self):
+        self._blobs = []
+        self._struct_parser = StructParser(HDR1_HEADER_DEFINITION)
+
+    def parse(self, file: File):
+        """Parse header + blob headers and populate self._blobs"""
+        header = self._struct_parser.parse("hdr_header_t", file, Endian.LITTLE)
+        for offset in header.blob_offsets:
+            if not offset:
+                continue
+            file.seek(offset, io.SEEK_SET)
+            blob_header = self._struct_parser.parse(
+                "blob_header_t", file, Endian.LITTLE
+            )
+            # file.tell() points to right after the blob_header
+            self.blobs.append(
+                Blob(
+                    offset=file.tell(),
+                    name=snull(blob_header.name).decode("utf-8"),
+                    size=blob_header.size,
+                )
+            )
+
+    @property
+    def blobs(self):
+        return self._blobs
+
+
+class HDRHandler(StructHandler):
+    NAME = "hdr1"
+
+    PATTERNS = [HexString("48 44 52 31")]
+
+    C_DEFINITIONS = HDR1_HEADER_DEFINITION
+    HEADER_STRUCT = "hdr_header_t"
+    EXTRACTOR = HDRExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        header = self.parse_header(file, endian=Endian.LITTLE)
+        end_offset = start_offset + header.signature_offset + SIGNATURE_LEN
```

Comment:
I would move the `end_offset` calculation to after the call to `is_valid_header`. This way we only compute it if it's a true positive.

Reply:
end_offset has been moved after the is_valid_header to avoid unnecessary calculation.